
syntax = "proto3";

option go_package = "pkg/cache";

package fstree;

message ReadObjectRequest {
    string digest = 1;
}

message ReadObjectResponse {
    bytes data = 1;
}

message WriteObjectRequest {
    string digest = 1;
    bytes data = 2;
}

message WriteObjectResponse {
}

message HasObjectRequest {
    repeated string digest = 1;
}

message HasObjectResponse {
    repeated bool present = 1;
}

message HasTreeRequest {
    repeated string digest = 1;
}

message HasTreeResponse {
    repeated string missing_trees = 1;
    repeated string missing_objects = 2;
}

service CacheService {
    // GetBlobPresence returns a list of blobs that are either missing or present in the cache.
    rpc HasObject(HasObjectRequest) returns (HasObjectResponse);

    // HasTree returns a list of objects and trees that are missing or present in the cache.
    rpc HasTree(HasTreeRequest) returns (stream HasTreeResponse);

    // Read a blob from the cache
    rpc ReadObject(ReadObjectRequest) returns (stream ReadObjectResponse);

    // Write a blob to the cache
    rpc WriteObject(stream WriteObjectRequest) returns (WriteObjectResponse);
}

